{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "import asyncio\n",
    "import sys\n",
    "sys.path.append('/Users/sebastian/Documents/Bachelor Thesis/sparql-rag-agent/sparql-rag-agent')\n",
    "from experiments.utilities.evaluation import AgentEvaluator"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# test_set = {\n",
    "#     \"SwissLipids\": [\"5.ttl\", \"34.ttl\"],\n",
    "#     \"Uniprot\": [\"51_all_proteins_linked_to_arachidonate.ttl\", \"40_human_enzymes_that_metabolize_sphingolipids\"],\n",
    "#     \"Rhea\": [\"12.ttl\", \"112_Total_number_of_reaction_ec_protein_links.ttl\", \"117_Retrieve_parent_reactions_and_their_catalyzing_enzymes_for_a_metabolite_that_is_not_directly_involved_in_a_reaction.ttl\"]\n",
    "# }\n",
    "\n",
    "test_set = { \"SwissLipids\": [], \"Uniprot\": [], \"Rhea\": [] }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing endpoint set: SwissLipids\n",
      "Processed: 16.ttl\n",
      "Processed: 6.ttl\n",
      "Processed: 5.ttl\n",
      "Processed: 34.ttl\n",
      "Processed 4 files from SwissLipids\n",
      "Processing endpoint set: Uniprot\n",
      "Processed: 118_biosodafrontend_rat_brain_human_cancer.ttl\n",
      "Processed: 51_all_proteins_linked_to_arachidonate.ttl\n",
      "Processed: 40_human_enzymes_that_metabolize_sphingolipids.ttl\n",
      "Processed: 49_tissues_where_genes_metabolizing_cholestrol_are_expressed.ttl\n",
      "Processed: 53_mouse_homologs_of_sterol_enzymes_via_omabrowser.ttl\n",
      "Processed: 52_drugs_targeting_sterol_metabolism.ttl\n",
      "Processed: 54_binding_sites_for_ligands_similar_to_heme_sachem.ttl\n",
      "Processed: 43_patents_used_in_uniprot_granted_twenty_years_in_the_past.ttl\n",
      "Processed: 45_drugs_targeting_human_sterol_metabolism_enzymes.ttl\n",
      "Processed: 42_connect_patents_to_epo.ttl\n",
      "Processed: 67_draft_human_metabolome.ttl\n",
      "Processed: 36_orthologous_proteins_via_orthodb.ttl\n",
      "Processed: 60_enzymes_working_on_substrate_with_Cholestene_backbone.ttl\n",
      "Processed: 70_enzymes_interacting_with_molecules_similar_to_dopamine.ttl\n",
      "Processed: 38_find_gene_region_in_wikidata_for_known_entry.ttl\n",
      "Processed: 117_biosodafrontend_glioblastoma_orthologs_rat.ttl\n",
      "Processed: 116_biosodafrontend_rabit_mouse_orthologs.ttl\n",
      "Processed: 29_search_uniprot_in_japanese_with_allie.ttl\n",
      "Processed 18 files from Uniprot\n",
      "Processing endpoint set: Rhea\n",
      "Processed: 94_Select_all_approved_reactions_with_CHEBI_or_one_of_its_descendant_optional_UniProtKB_reviewed_EC.ttl\n",
      "Processed: 11.ttl\n",
      "Processed: 109_Distribution_of_reactions_according_to_the_first_level_of_enzyme_classification.ttl\n",
      "Processed: 12.ttl\n",
      "Processed: 112_Total_number_of_reaction_ec_protein_links.ttl\n",
      "Processed: 19_draft_human_metabolome.ttl\n",
      "Processed: 20_search_chemical_names_in_japanese.ttl\n",
      "Processed: 117_Retrieve_parent_reactions_and_their_catalyzing_enzymes_for_a_metabolite_that_is_not_directly_involved_in_a_reaction.ttl\n",
      "Processed: 18.ttl\n",
      "Processed: 115_Retrieve_the_set_of_metabolites_involved_in_reactions_catalyzed_by_human_genes_HUMAN_reference_proteome.ttl\n",
      "Processed 10 files from Rhea\n",
      "Saved metadata for 32 files from 3 endpoint sets to:\n",
      "  - /Users/sebastian/Documents/Bachelor Thesis/sparql-rag-agent/sparql-rag-agent/experiments/experiments_official/ev_2025-04-22_02-04-22/testset_meta_data.json\n",
      "Sending question 1/32 to agent...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING: All log messages before absl::InitializeLog() is called are written to STDERR\n",
      "I0000 00:00:1745280265.841191  725053 fork_posix.cc:75] Other threads are currently calling into gRPC, skipping fork() handlers\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Warning: Run has None value for start_time or end_time\n",
      "Got result for question 1/32\n",
      "Querying ground truth endpoint...\n",
      "  Using cached result for query\n",
      "Querying predicted endpoint...\n",
      "Sending question 2/32 to agent...\n",
      "Warning: Run has None value for start_time or end_time\n",
      "Got result for question 2/32\n",
      "Sending question 3/32 to agent...\n",
      "Warning: Run has None value for start_time or end_time\n",
      "Got result for question 3/32\n",
      "Querying ground truth endpoint...\n",
      "  Using cached result for query\n",
      "Querying predicted endpoint...\n",
      "Sending question 4/32 to agent...\n",
      "Warning: Run has None value for start_time or end_time\n",
      "Got result for question 4/32\n",
      "Querying ground truth endpoint...\n",
      "  Using cached result for query\n",
      "Querying predicted endpoint...\n",
      "Sending question 5/32 to agent...\n",
      "Warning: Run has None value for start_time or end_time\n",
      "Got result for question 5/32\n",
      "Sending question 6/32 to agent...\n",
      "Warning: Run has None value for start_time or end_time\n",
      "Got result for question 6/32\n",
      "Querying ground truth endpoint...\n",
      "  Using cached result for query\n",
      "Querying predicted endpoint...\n",
      "Sending question 7/32 to agent...\n",
      "Warning: Run has None value for start_time or end_time\n",
      "Got result for question 7/32\n",
      "Querying ground truth endpoint...\n",
      "  Using cached result for query\n",
      "Querying predicted endpoint...\n",
      "Sending question 8/32 to agent...\n",
      "Warning: Run has None value for start_time or end_time\n",
      "Got result for question 8/32\n",
      "Querying ground truth endpoint...\n",
      "  Using cached result for query\n",
      "Querying predicted endpoint...\n",
      "Sending question 9/32 to agent...\n",
      "Warning: Run has None value for start_time or end_time\n",
      "Got result for question 9/32\n",
      "Querying ground truth endpoint...\n",
      "  Using cached result for query\n",
      "Querying predicted endpoint...\n",
      "Sending question 10/32 to agent...\n",
      "Warning: Run has None value for start_time or end_time\n",
      "Got result for question 10/32\n",
      "Querying ground truth endpoint...\n",
      "  Using cached result for query\n",
      "Querying predicted endpoint...\n",
      "Sending question 11/32 to agent...\n",
      "Warning: Run has None value for start_time or end_time\n",
      "Got result for question 11/32\n",
      "Sending question 12/32 to agent...\n",
      "Warning: Run has None value for start_time or end_time\n",
      "Got result for question 12/32\n",
      "Querying ground truth endpoint...\n",
      "  Using cached result for query\n",
      "Querying predicted endpoint...\n",
      "Sending question 13/32 to agent...\n",
      "Warning: Run has None value for start_time or end_time\n",
      "Got result for question 13/32\n",
      "Sending question 14/32 to agent...\n",
      "Warning: Run has None value for start_time or end_time\n",
      "Got result for question 14/32\n",
      "Querying ground truth endpoint...\n",
      "  Using cached result for query\n",
      "Querying predicted endpoint...\n",
      "Sending question 15/32 to agent...\n",
      "Warning: Run has None value for start_time or end_time\n",
      "Got result for question 15/32\n",
      "Sending question 16/32 to agent...\n",
      "Warning: Run has None value for start_time or end_time\n",
      "Got result for question 16/32\n",
      "Querying ground truth endpoint...\n",
      "  Using cached result for query\n",
      "Querying predicted endpoint...\n",
      "Sending question 17/32 to agent...\n",
      "Warning: Run has None value for start_time or end_time\n",
      "Got result for question 17/32\n",
      "Querying ground truth endpoint...\n",
      "  Using cached result for query\n",
      "Querying predicted endpoint...\n",
      "Sending question 18/32 to agent...\n",
      "Warning: Run has None value for start_time or end_time\n",
      "Got result for question 18/32\n",
      "Sending question 19/32 to agent...\n",
      "Warning: Run has None value for start_time or end_time\n",
      "Got result for question 19/32\n",
      "Querying ground truth endpoint...\n",
      "  Using cached result for query\n",
      "Querying predicted endpoint...\n",
      "Sending question 20/32 to agent...\n",
      "Warning: Run has None value for start_time or end_time\n",
      "Got result for question 20/32\n",
      "Sending question 21/32 to agent...\n",
      "Warning: Run has None value for start_time or end_time\n",
      "Got result for question 21/32\n",
      "Sending question 22/32 to agent...\n",
      "Warning: Run has None value for start_time or end_time\n",
      "Got result for question 22/32\n",
      "Querying ground truth endpoint...\n",
      "  Using cached result for query\n",
      "Querying predicted endpoint...\n",
      "Sending question 23/32 to agent...\n",
      "Warning: Run has None value for start_time or end_time\n",
      "Got result for question 23/32\n",
      "Querying ground truth endpoint...\n",
      "  Using cached result for query\n",
      "Querying predicted endpoint...\n",
      "Sending question 24/32 to agent...\n",
      "Warning: Run has None value for start_time or end_time\n",
      "Got result for question 24/32\n",
      "Querying ground truth endpoint...\n",
      "  Using cached result for query\n",
      "Querying predicted endpoint...\n",
      "Sending question 25/32 to agent...\n",
      "Warning: Run has None value for start_time or end_time\n",
      "Got result for question 25/32\n",
      "Querying ground truth endpoint...\n",
      "  Using cached result for query\n",
      "Querying predicted endpoint...\n",
      "Sending question 26/32 to agent...\n",
      "Warning: Run has None value for start_time or end_time\n",
      "Got result for question 26/32\n",
      "Querying ground truth endpoint...\n",
      "  Using cached result for query\n",
      "Querying predicted endpoint...\n",
      "Sending question 27/32 to agent...\n",
      "Warning: Run has None value for start_time or end_time\n",
      "Got result for question 27/32\n",
      "Querying ground truth endpoint...\n",
      "  Using cached result for query\n",
      "Querying predicted endpoint...\n",
      "Sending question 28/32 to agent...\n",
      "Warning: Run has None value for start_time or end_time\n",
      "Got result for question 28/32\n",
      "Querying ground truth endpoint...\n",
      "  Using cached result for query\n",
      "Querying predicted endpoint...\n",
      "Sending question 29/32 to agent...\n",
      "Warning: Run has None value for start_time or end_time\n",
      "Got result for question 29/32\n",
      "Querying ground truth endpoint...\n",
      "  Using cached result for query\n",
      "Querying predicted endpoint...\n",
      "Sending question 30/32 to agent...\n",
      "Warning: Run has None value for start_time or end_time\n",
      "Got result for question 30/32\n",
      "Querying ground truth endpoint...\n",
      "  Using cached result for query\n",
      "Querying predicted endpoint...\n",
      "Sending question 31/32 to agent...\n",
      "Warning: Run has None value for start_time or end_time\n",
      "Got result for question 31/32\n",
      "Querying ground truth endpoint...\n",
      "  Using cached result for query\n",
      "Querying predicted endpoint...\n",
      "Sending question 32/32 to agent...\n",
      "Warning: Run has None value for start_time or end_time\n",
      "Got result for question 32/32\n",
      "Querying ground truth endpoint...\n",
      "  Using cached result for query\n",
      "Querying predicted endpoint...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/sebastian/Documents/Bachelor Thesis/sparql-rag-agent/sparql-rag-agent/.venv/lib/python3.11/site-packages/nltk/translate/bleu_score.py:577: UserWarning: \n",
      "The hypothesis contains 0 counts of 4-gram overlaps.\n",
      "Therefore the BLEU score evaluates to 0, independently of\n",
      "how many N-gram overlaps of lower order it contains.\n",
      "Consider using lower n-gram order or use SmoothingFunction()\n",
      "  warnings.warn(_msg)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "None query found\n",
      "None query found\n",
      "None query found\n",
      "None query found\n",
      "None query found\n",
      "SP-BLEU 0.028329313952968147\n",
      "METEOR 0.27928359891062843\n",
      "F1-score:  0\n",
      "SP-F1-score:  0\n",
      "Number of invalid queries(None): 5\n",
      "Saved query comparison to: /Users/sebastian/Documents/Bachelor Thesis/sparql-rag-agent/sparql-rag-agent/experiments/experiments_official/ev_2025-04-22_02-04-22/16_comparison.ttl\n",
      "Saved query comparison to: /Users/sebastian/Documents/Bachelor Thesis/sparql-rag-agent/sparql-rag-agent/experiments/experiments_official/ev_2025-04-22_02-04-22/6_comparison.ttl\n",
      "Saved query comparison to: /Users/sebastian/Documents/Bachelor Thesis/sparql-rag-agent/sparql-rag-agent/experiments/experiments_official/ev_2025-04-22_02-04-22/5_comparison.ttl\n",
      "Saved query comparison to: /Users/sebastian/Documents/Bachelor Thesis/sparql-rag-agent/sparql-rag-agent/experiments/experiments_official/ev_2025-04-22_02-04-22/34_comparison.ttl\n",
      "Saved query comparison to: /Users/sebastian/Documents/Bachelor Thesis/sparql-rag-agent/sparql-rag-agent/experiments/experiments_official/ev_2025-04-22_02-04-22/118_biosodafrontend_rat_brain_human_cancer_comparison.ttl\n",
      "Saved query comparison to: /Users/sebastian/Documents/Bachelor Thesis/sparql-rag-agent/sparql-rag-agent/experiments/experiments_official/ev_2025-04-22_02-04-22/51_comparison.ttl\n",
      "Saved query comparison to: /Users/sebastian/Documents/Bachelor Thesis/sparql-rag-agent/sparql-rag-agent/experiments/experiments_official/ev_2025-04-22_02-04-22/40_comparison.ttl\n",
      "Saved query comparison to: /Users/sebastian/Documents/Bachelor Thesis/sparql-rag-agent/sparql-rag-agent/experiments/experiments_official/ev_2025-04-22_02-04-22/49_comparison.ttl\n",
      "Saved query comparison to: /Users/sebastian/Documents/Bachelor Thesis/sparql-rag-agent/sparql-rag-agent/experiments/experiments_official/ev_2025-04-22_02-04-22/53_comparison.ttl\n",
      "Saved query comparison to: /Users/sebastian/Documents/Bachelor Thesis/sparql-rag-agent/sparql-rag-agent/experiments/experiments_official/ev_2025-04-22_02-04-22/52_comparison.ttl\n",
      "Saved query comparison to: /Users/sebastian/Documents/Bachelor Thesis/sparql-rag-agent/sparql-rag-agent/experiments/experiments_official/ev_2025-04-22_02-04-22/54_comparison.ttl\n",
      "Saved query comparison to: /Users/sebastian/Documents/Bachelor Thesis/sparql-rag-agent/sparql-rag-agent/experiments/experiments_official/ev_2025-04-22_02-04-22/43_comparison.ttl\n",
      "Saved query comparison to: /Users/sebastian/Documents/Bachelor Thesis/sparql-rag-agent/sparql-rag-agent/experiments/experiments_official/ev_2025-04-22_02-04-22/45_comparison.ttl\n",
      "Saved query comparison to: /Users/sebastian/Documents/Bachelor Thesis/sparql-rag-agent/sparql-rag-agent/experiments/experiments_official/ev_2025-04-22_02-04-22/42_comparison.ttl\n",
      "Saved query comparison to: /Users/sebastian/Documents/Bachelor Thesis/sparql-rag-agent/sparql-rag-agent/experiments/experiments_official/ev_2025-04-22_02-04-22/67_draft_human_metabolome_comparison.ttl\n",
      "Saved query comparison to: /Users/sebastian/Documents/Bachelor Thesis/sparql-rag-agent/sparql-rag-agent/experiments/experiments_official/ev_2025-04-22_02-04-22/36_comparison.ttl\n",
      "Saved query comparison to: /Users/sebastian/Documents/Bachelor Thesis/sparql-rag-agent/sparql-rag-agent/experiments/experiments_official/ev_2025-04-22_02-04-22/60_comparison.ttl\n",
      "Saved query comparison to: /Users/sebastian/Documents/Bachelor Thesis/sparql-rag-agent/sparql-rag-agent/experiments/experiments_official/ev_2025-04-22_02-04-22/70_enzymes_interacting_with_molecules_similar_to_dopamine_comparison.ttl\n",
      "Saved query comparison to: /Users/sebastian/Documents/Bachelor Thesis/sparql-rag-agent/sparql-rag-agent/experiments/experiments_official/ev_2025-04-22_02-04-22/38_comparison.ttl\n",
      "Saved query comparison to: /Users/sebastian/Documents/Bachelor Thesis/sparql-rag-agent/sparql-rag-agent/experiments/experiments_official/ev_2025-04-22_02-04-22/117_biosodafrontend_glioblastoma_orthologs_rat_comparison.ttl\n",
      "Saved query comparison to: /Users/sebastian/Documents/Bachelor Thesis/sparql-rag-agent/sparql-rag-agent/experiments/experiments_official/ev_2025-04-22_02-04-22/116_biosodafrontend_rabit_mouse_orthologs_comparison.ttl\n",
      "Saved query comparison to: /Users/sebastian/Documents/Bachelor Thesis/sparql-rag-agent/sparql-rag-agent/experiments/experiments_official/ev_2025-04-22_02-04-22/29_comparison.ttl\n",
      "Saved query comparison to: /Users/sebastian/Documents/Bachelor Thesis/sparql-rag-agent/sparql-rag-agent/experiments/experiments_official/ev_2025-04-22_02-04-22/94_Select_all_approved_reactions_with_CHEBI_or_one_of_its_descendant_optional_UniProtKB_reviewed_EC_comparison.ttl\n",
      "Saved query comparison to: /Users/sebastian/Documents/Bachelor Thesis/sparql-rag-agent/sparql-rag-agent/experiments/experiments_official/ev_2025-04-22_02-04-22/11_comparison.ttl\n",
      "Saved query comparison to: /Users/sebastian/Documents/Bachelor Thesis/sparql-rag-agent/sparql-rag-agent/experiments/experiments_official/ev_2025-04-22_02-04-22/109_Distribution_of_reactions_according_to_the_first_level_of_enzyme_classification_comparison.ttl\n",
      "Saved query comparison to: /Users/sebastian/Documents/Bachelor Thesis/sparql-rag-agent/sparql-rag-agent/experiments/experiments_official/ev_2025-04-22_02-04-22/12_comparison.ttl\n",
      "Saved query comparison to: /Users/sebastian/Documents/Bachelor Thesis/sparql-rag-agent/sparql-rag-agent/experiments/experiments_official/ev_2025-04-22_02-04-22/112_Total_number_of_reaction_ec_protein_links_comparison.ttl\n",
      "Saved query comparison to: /Users/sebastian/Documents/Bachelor Thesis/sparql-rag-agent/sparql-rag-agent/experiments/experiments_official/ev_2025-04-22_02-04-22/19_draft_human_metabolome_comparison.ttl\n",
      "Saved query comparison to: /Users/sebastian/Documents/Bachelor Thesis/sparql-rag-agent/sparql-rag-agent/experiments/experiments_official/ev_2025-04-22_02-04-22/20_search_chemical_names_in_japanese_comparison.ttl\n",
      "Saved query comparison to: /Users/sebastian/Documents/Bachelor Thesis/sparql-rag-agent/sparql-rag-agent/experiments/experiments_official/ev_2025-04-22_02-04-22/117_Retrieve_parent_reactions_and_their_catalyzing_enzymes_for_a_metabolite_that_is_not_directly_involved_in_a_reaction_comparison.ttl\n",
      "Saved query comparison to: /Users/sebastian/Documents/Bachelor Thesis/sparql-rag-agent/sparql-rag-agent/experiments/experiments_official/ev_2025-04-22_02-04-22/18_comparison.ttl\n",
      "Saved query comparison to: /Users/sebastian/Documents/Bachelor Thesis/sparql-rag-agent/sparql-rag-agent/experiments/experiments_official/ev_2025-04-22_02-04-22/115_Retrieve_the_set_of_metabolites_involved_in_reactions_catalyzed_by_human_genes_HUMAN_reference_proteome_comparison.ttl\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/sebastian/Documents/Bachelor Thesis/sparql-rag-agent/sparql-rag-agent/.venv/lib/python3.11/site-packages/nltk/translate/bleu_score.py:577: UserWarning: \n",
      "The hypothesis contains 0 counts of 3-gram overlaps.\n",
      "Therefore the BLEU score evaluates to 0, independently of\n",
      "how many N-gram overlaps of lower order it contains.\n",
      "Consider using lower n-gram order or use SmoothingFunction()\n",
      "  warnings.warn(_msg)\n"
     ]
    }
   ],
   "source": [
    "\n",
    "evaluator = AgentEvaluator(\n",
    "    endpoint_sets=test_set,\n",
    "    experiment_dir=\"experiments_official\",\n",
    "    project_name_langsmith=\"sparql-rag-agent-construction-prompt-version-1.2\",\n",
    "    timeout = 300,\n",
    "    dataset_dir=\"/Users/sebastian/Documents/Bachelor Thesis/sparql-rag-agent/sparql-rag-agent/experiments/federated_sparql_dataset/examples_federated_19.04.2025\"\n",
    "\n",
    ")\n",
    "evaluation_dataset_path = await evaluator.run_all_tests()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "\n",
    "evaluation_dataset_path = \"/Users/sebastian/Documents/Bachelor Thesis/sparql-rag-agent/sparql-rag-agent/experiments/experiments_official/ev_2025-04-16_11-21-33/evaluation_dataset.json\"\n",
    "df = pd.read_json(evaluation_dataset_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>resource</th>\n",
       "      <th>natural_language_question</th>\n",
       "      <th>ground_truth_query</th>\n",
       "      <th>target_endpoint</th>\n",
       "      <th>federates_with</th>\n",
       "      <th>endpoint_set</th>\n",
       "      <th>file_path</th>\n",
       "      <th>predicted_query</th>\n",
       "      <th>predicted_endpoint</th>\n",
       "      <th>predicted_endpoint_equal_to_target_endpoint</th>\n",
       "      <th>...</th>\n",
       "      <th>sparql_construction_completion_tokens</th>\n",
       "      <th>sparql_construction_total_tokens</th>\n",
       "      <th>is_valid_sparql</th>\n",
       "      <th>sparql_syntax_error</th>\n",
       "      <th>ground_truth_query_result_is_empty</th>\n",
       "      <th>result_eval_f1_score</th>\n",
       "      <th>result_eval_precision</th>\n",
       "      <th>result_eval_recall</th>\n",
       "      <th>error_occured_at_endpoint</th>\n",
       "      <th>predicted_query_result_is_empty</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>https://sparql.swisslipids.org/.well-known/spa...</td>\n",
       "      <td>For a given lipid (SWISSLIPID:000399814, Ceram...</td>\n",
       "      <td>PREFIX SWISSLIPID: &lt;https://swisslipids.org/rd...</td>\n",
       "      <td>https://sparql.swisslipids.org/sparql/</td>\n",
       "      <td>[https://sparql.uniprot.org/sparql, https://sp...</td>\n",
       "      <td>SwissLipids</td>\n",
       "      <td>SwissLipids/6.ttl</td>\n",
       "      <td># https://sparql.swisslipids.ch/sparql\\nPREFIX...</td>\n",
       "      <td>https://sparql.swisslipids.ch/sparql</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>2361</td>\n",
       "      <td>False</td>\n",
       "      <td>Line 11, column 12: Unresolved prefixed name: ...</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>https://sparql.swisslipids.org/.well-known/spa...</td>\n",
       "      <td>For a given lipid (SWISSLIPID:000399814, Ceram...</td>\n",
       "      <td>PREFIX SWISSLIPID: &lt;https://swisslipids.org/rd...</td>\n",
       "      <td>https://sparql.swisslipids.org/sparql/</td>\n",
       "      <td>[https://sparql.rhea-db.org/sparql]</td>\n",
       "      <td>SwissLipids</td>\n",
       "      <td>SwissLipids/5.ttl</td>\n",
       "      <td># https://sparql.rhea-db.org/sparql\\nPREFIX rh...</td>\n",
       "      <td>https://sparql.rhea-db.org/sparql</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1669</td>\n",
       "      <td>False</td>\n",
       "      <td>Line 13, column 36: Unresolved prefixed name: ...</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>https://sparql.swisslipids.org/.well-known/spa...</td>\n",
       "      <td>Lipids affected by human enzymes</td>\n",
       "      <td>PREFIX owl: &lt;http://www.w3.org/2002/07/owl#&gt;\\n...</td>\n",
       "      <td>https://sparql.swisslipids.org/sparql/</td>\n",
       "      <td>[https://sparql.uniprot.org/sparql, https://sp...</td>\n",
       "      <td>SwissLipids</td>\n",
       "      <td>SwissLipids/34.ttl</td>\n",
       "      <td># https://sparql.uniprot.org/sparql\\nPREFIX up...</td>\n",
       "      <td>https://sparql.uniprot.org/sparql</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>1086</td>\n",
       "      <td>2328</td>\n",
       "      <td>False</td>\n",
       "      <td>Line 27, column 12: Unresolved prefixed name: ...</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>https://sparql.uniprot.org/.well-known/sparql-...</td>\n",
       "      <td>What are the Homo sapiens genes associated wit...</td>\n",
       "      <td>PREFIX up:&lt;http://purl.uniprot.org/core/&gt;\\nPRE...</td>\n",
       "      <td>https://sparql.uniprot.org/sparql</td>\n",
       "      <td>[https://sparql.omabrowser.org/sparql/, https:...</td>\n",
       "      <td>Uniprot</td>\n",
       "      <td>UniProt/118_biosodafrontend_rat_brain_human_ca...</td>\n",
       "      <td># https://sparql.uniprot.org/sparql\\nPREFIX ta...</td>\n",
       "      <td>https://sparql.uniprot.org/sparql</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>481</td>\n",
       "      <td>2435</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>https://sparql.uniprot.org/.well-known/sparql-...</td>\n",
       "      <td>Find all proteins linked to arachidonate (CHEB...</td>\n",
       "      <td>PREFIX CHEBI: &lt;http://purl.obolibrary.org/obo/...</td>\n",
       "      <td>https://sparql.uniprot.org/sparql/</td>\n",
       "      <td>[https://sparql.rhea-db.org/sparql]</td>\n",
       "      <td>Uniprot</td>\n",
       "      <td>UniProt/51_all_proteins_linked_to_arachidonate...</td>\n",
       "      <td># https://sparql.uniprot.org/sparql\\nPREFIX up...</td>\n",
       "      <td>https://sparql.uniprot.org/sparql</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1537</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 27 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                            resource  \\\n",
       "0  https://sparql.swisslipids.org/.well-known/spa...   \n",
       "1  https://sparql.swisslipids.org/.well-known/spa...   \n",
       "2  https://sparql.swisslipids.org/.well-known/spa...   \n",
       "3  https://sparql.uniprot.org/.well-known/sparql-...   \n",
       "4  https://sparql.uniprot.org/.well-known/sparql-...   \n",
       "\n",
       "                           natural_language_question  \\\n",
       "0  For a given lipid (SWISSLIPID:000399814, Ceram...   \n",
       "1  For a given lipid (SWISSLIPID:000399814, Ceram...   \n",
       "2                   Lipids affected by human enzymes   \n",
       "3  What are the Homo sapiens genes associated wit...   \n",
       "4  Find all proteins linked to arachidonate (CHEB...   \n",
       "\n",
       "                                  ground_truth_query  \\\n",
       "0  PREFIX SWISSLIPID: <https://swisslipids.org/rd...   \n",
       "1  PREFIX SWISSLIPID: <https://swisslipids.org/rd...   \n",
       "2  PREFIX owl: <http://www.w3.org/2002/07/owl#>\\n...   \n",
       "3  PREFIX up:<http://purl.uniprot.org/core/>\\nPRE...   \n",
       "4  PREFIX CHEBI: <http://purl.obolibrary.org/obo/...   \n",
       "\n",
       "                          target_endpoint  \\\n",
       "0  https://sparql.swisslipids.org/sparql/   \n",
       "1  https://sparql.swisslipids.org/sparql/   \n",
       "2  https://sparql.swisslipids.org/sparql/   \n",
       "3       https://sparql.uniprot.org/sparql   \n",
       "4      https://sparql.uniprot.org/sparql/   \n",
       "\n",
       "                                      federates_with endpoint_set  \\\n",
       "0  [https://sparql.uniprot.org/sparql, https://sp...  SwissLipids   \n",
       "1                [https://sparql.rhea-db.org/sparql]  SwissLipids   \n",
       "2  [https://sparql.uniprot.org/sparql, https://sp...  SwissLipids   \n",
       "3  [https://sparql.omabrowser.org/sparql/, https:...      Uniprot   \n",
       "4                [https://sparql.rhea-db.org/sparql]      Uniprot   \n",
       "\n",
       "                                           file_path  \\\n",
       "0                                  SwissLipids/6.ttl   \n",
       "1                                  SwissLipids/5.ttl   \n",
       "2                                 SwissLipids/34.ttl   \n",
       "3  UniProt/118_biosodafrontend_rat_brain_human_ca...   \n",
       "4  UniProt/51_all_proteins_linked_to_arachidonate...   \n",
       "\n",
       "                                     predicted_query  \\\n",
       "0  # https://sparql.swisslipids.ch/sparql\\nPREFIX...   \n",
       "1  # https://sparql.rhea-db.org/sparql\\nPREFIX rh...   \n",
       "2  # https://sparql.uniprot.org/sparql\\nPREFIX up...   \n",
       "3  # https://sparql.uniprot.org/sparql\\nPREFIX ta...   \n",
       "4  # https://sparql.uniprot.org/sparql\\nPREFIX up...   \n",
       "\n",
       "                     predicted_endpoint  \\\n",
       "0  https://sparql.swisslipids.ch/sparql   \n",
       "1     https://sparql.rhea-db.org/sparql   \n",
       "2     https://sparql.uniprot.org/sparql   \n",
       "3     https://sparql.uniprot.org/sparql   \n",
       "4     https://sparql.uniprot.org/sparql   \n",
       "\n",
       "   predicted_endpoint_equal_to_target_endpoint  ...  \\\n",
       "0                                        False  ...   \n",
       "1                                        False  ...   \n",
       "2                                        False  ...   \n",
       "3                                         True  ...   \n",
       "4                                         True  ...   \n",
       "\n",
       "   sparql_construction_completion_tokens sparql_construction_total_tokens  \\\n",
       "0                                      0                             2361   \n",
       "1                                      0                             1669   \n",
       "2                                   1086                             2328   \n",
       "3                                    481                             2435   \n",
       "4                                      0                             1537   \n",
       "\n",
       "  is_valid_sparql                                sparql_syntax_error  \\\n",
       "0           False  Line 11, column 12: Unresolved prefixed name: ...   \n",
       "1           False  Line 13, column 36: Unresolved prefixed name: ...   \n",
       "2           False  Line 27, column 12: Unresolved prefixed name: ...   \n",
       "3            True                                               None   \n",
       "4            True                                               None   \n",
       "\n",
       "   ground_truth_query_result_is_empty  result_eval_f1_score  \\\n",
       "0                               False                     0   \n",
       "1                               False                     0   \n",
       "2                               False                     0   \n",
       "3                               False                     0   \n",
       "4                               False                     0   \n",
       "\n",
       "   result_eval_precision  result_eval_recall  error_occured_at_endpoint  \\\n",
       "0                      0                   0                        1.0   \n",
       "1                      0                   0                        0.0   \n",
       "2                      0                   0                        1.0   \n",
       "3                      0                   0                        0.0   \n",
       "4                      0                   0                        0.0   \n",
       "\n",
       "   predicted_query_result_is_empty  \n",
       "0                             True  \n",
       "1                             True  \n",
       "2                             True  \n",
       "3                             True  \n",
       "4                             True  \n",
       "\n",
       "[5 rows x 27 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "29\n"
     ]
    }
   ],
   "source": [
    "print(len(df))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "is_valid_sparql\n",
      "True     21\n",
      "False     8\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "print(df['is_valid_sparql'].value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "error_occured_at_endpoint\n",
      "0.0    16\n",
      "1.0    10\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "print(df['error_occured_at_endpoint'].value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "predicted_endpoint_equal_to_target_endpoint\n",
      "True     17\n",
      "False    12\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "print(df['predicted_endpoint_equal_to_target_endpoint'].value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "predicted_endpoint_in_federates_with\n",
      "False    23\n",
      "True      6\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "print(df['predicted_endpoint_in_federates_with'].value_counts())"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
